Feature: Documentation (#62)
MomirMilutinovic/graph-expressiveness-team-6: README.md

#### 3. Edge Model

The `Edge` model represents an edge between two nodes in the graph.
Edges, like nodes, encapsulate edge-specific properties and data.

#### 4. Plugin Abstractions

Application supports two types of plugins: `data source plugins` and `visualizer plugins`.

The API defines two abstract classes:

- `Visualizer`: Abstract class for implementing Visualizer plugins. The structure is as follows:
  ```python
  class Visualizer(ABC):
      @abstractmethod
      def get_name(self) -> str:
          pass
  
      @abstractmethod
      def display(self, graph: Graph):
          pass
  ```
- `DataSource`: Abstract class for implementing Data Source plugins. The structure is as follows:
  ```python
  class DataSource(ABC):
      @abstractmethod
      def get_name(self) -> str:
          pass
  
      @abstractmethod
      def provide(self, **kwargs) -> Graph:
          pass
  
      @abstractmethod
      def get_configuration_parameters(self) -> dict[str, str]:
          pass
  ```

These abstract classes provide a foundation for creating plugins by specifying required methods and properties.

## Plugin Development

> In order to create a new plugin, you need to implement the interface similar as API package abstraction and provide an entry point for the plugin.

Reviewer: ```suggestion
In order to create a new plugin, you need to implement the interface for the kind of plugin you are writing and provide an entry point for the plugin.
```

   - Source Code: [Advanced Visualizer Plugin](./advanced_visualizer)

### Software Components

The software is divided into three main views:
- **Main View**
- **Tree View**
- **Bird View**

### Main View
`The Main View` represents a canvas that visualizes the graph in two ways using `visualizer` plugins.

Provides operations such as `pan`, `zoom`, `drag` and `drop`
- Nodes are draggable and can be moved around the canvas. 
- The canvas can be panned and zoomed in and out using the mouse or touchpad.

### Tree View
`The Tree View` (graph representation as a tree) allows dynamic expanding and collapsing of nodes,
similar to the package explorer in IDEs.

- `plus` sign next to an object indicates that it can be expanded
- `minus` sign means that the object can be collapsed.

## Bird View

`Bird View` (Graph Overview from a Bird's-Eye Perspective) is a reduced representation of the graph that must entirely fit within the canvas.

Bird View mirrors the Main View with the key difference that the entire graph fits into the canvas. 
Bird View encompass the viewport that moves in accordance with the current display on the Main View.
The behavior of the viewport is defined as follows:

> - The `zoom-out` on the Main View reducing the viewport on the Bird View.

Reviewer: ```suggestion
- Zooming out the Main View shrinks the viewport on the Bird View.
```

> - The `zoom-in` on the Main View enlarging the viewport on the Bird View.

Reviewer: ```suggestion
- Zooming in the Main View enlarges the viewport on the Bird View.
```

> - The `pan` operation on the Main View moving the viewport on the Bird View.

Reviewer: ```suggestion
- Panning the Main View moves the viewport on the Bird View.
```

> - `Dragging` the nodes on the Main View affecting the graph on the Bird View.

Reviewer: ```suggestion
- `Dragging` the nodes on the Main View affects the graph on the Bird View.
```

> All three graph views are simultaneously accessible and present the same graph in described ways.

Reviewer: ```suggestion
All three graph views are simultaneously accessible and present the same graph in the aforementioned ways.
```

> ### Graph View Interactions

Reviewer: Would feature be a better word for this?
```suggestion
### Graph View Features
```

- **Main View:**

>   - Pan: Allows dragging the graph to view different parts.

Reviewer: ```suggestion
  - Pan: Allows dragging the graph to view different parts of it.
```

>   - Zoom In/Out: Enables zooming in and out of the graph.

Reviewer: ```suggestion
  - Zoom In/Out: The graph can be zoomed in or out.
```
